packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m

```diff
@@ -675,99 +700,116 @@ - (NSDictionary *)encodingMap
-    NSData *data = [NSData dataWithContentsOfURL:scopedLocation];
-    if (!data) {
-      reject(@"E_UNABLE_TO_SAVE",
-             @"Unable to save file to local URI",
-             nil);
-      return;
-    }
-    NSMutableDictionary *result = [NSMutableDictionary dictionary];
-    result[@"uri"] = scopedLocation.absoluteString;
-    result[@"complete"] = @(YES);
-    if (options[@"md5"]) {
-      result[@"md5"] = [data md5String];
-    }
-    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)task.response;
-    result[@"status"] = @([httpResponse statusCode]);
-    result[@"headers"] = [httpResponse allHeaderFields];
-    
-    __strong EXFileSystem *strongSelf = weakSelf;
-    if (strongSelf) {
-      [strongSelf.downloadObjects removeObjectForKey:uuid];
+  return request;
+}
+
+- (EXFileSystemSessionType)_importSessionType:(NSNumber * _Nullable)type
+{
+  return [type intValue] ?: EXFileSystemBackgroundSession;
+}
+
+- (NSURLSession *)_sessionForType:(EXFileSystemSessionType)type
+{
+  switch (type) {
+    case EXFileSystemBackgroundSession:
+      return _backgroundSession;
+    case EXFileSystemForegroundSession:
+      return _foregroundSession;
+  }
+  return nil;
+}
+
+- (BOOL)_checkHeadersDictionary:(NSDictionary * _Nullable)headers
+{
+  for (id key in [headers allKeys]) {
+    if (![key isKindOfClass:[NSString class]] || ![headers[key] isKindOfClass:[NSString class]]) {
+      return false;
     }
-    
-    resolve(result);
-  };
+  }
   
-  EXDownloadDelegateOnErrorCallback onError = ^(NSError *error) {
-    //"cancelled" description when paused.  Don't throw.
-    if ([error.localizedDescription isEqualToString:@"cancelled"]) {
-      __strong EXFileSystem *strongSelf = weakSelf;
-      if (strongSelf) {
-        [strongSelf.downloadObjects removeObjectForKey:uuid];
-      }
-      resolve([NSNull null]);
-    } else {
-      reject(@"E_UNABLE_TO_DOWNLOAD",
-             [NSString stringWithFormat:@"Unable to download from: %@", url.absoluteString],
-             error);
-    }
+  return true;
+}
+
+- (NSURLSession *)_createSession:(EXFileSystemSessionType)type
+{
+  NSURLSessionConfiguration *sessionConfiguration;
+  if (type == EXFileSystemForegroundSession) {
+    sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+  } else {
+    sessionConfiguration = [NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:[[NSUUID UUID] UUIDString]];
+  }
+  sessionConfiguration.requestCachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
+  sessionConfiguration.URLCache = nil;
+  return [NSURLSession sessionWithConfiguration:sessionConfiguration
+                                       delegate:_sessionTaskDispatcher
+                                  delegateQueue:nil];
+}
+
+- (BOOL)_checkIfFileExists:(NSString *)path
+{
+  return [[NSFileManager defaultManager] fileExistsAtPath:path];
+}
+
+- (NSString * _Nullable)_importHttpMethod:(NSNumber *)httpMethod
```

Comment:
I've commented on that here: https://github.com/expo/expo/pull/7380#discussion_r406295401 ;) 

Moreover, I don't like strings here, cause someone might pass `get` - that will crash the app. So,  it would be nice to have some kind of validations. And what if someone misspelled it.

Reply:
IMHO it would be best to:
- `try-catch` not to crash the app if a user passed a `get`
- throw an error with a helpful message if an invalid method has been passed in.

This would make sure we don't limit the developers if they wanted to use some other HTTP method that is supported we don't know of. But I don't have a very strong opinion, so I will defer to @bbarthec's opinion on that.

Reply:
@lukmccall, I've discussed the thing with @sjchmiela and I think that going with `fetch-like` approach is desirable: https://fetch.spec.whatwg.org/#methods
I suggest enforcing TS type: `type AcceptedHttpMethod = 'POST' | 'PUT' | 'PATCH'` is sufficient guarding point, but later on let's go with `String.toUpperCase()` anyway while passing arguments to native side.
Are there any consequences of passing method that is not expected, eg. `CHICKEN` (I mean: are we semantically base on this value?)

Reply:
> Are there any consequences of passing method that is not expected, eg. CHICKEN (I mean: are we semantically base on this value?)

The task will be canceled by the system later on. Then we will reject the promise. So, I think that everything will work fine ;)